basis/cli/services/login.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+import base64
+import dataclasses
+import hashlib
+import os
+import re
+import urllib.parse
+from socketserver import BaseRequestHandler
+from urllib.parse import ParseResult
+
+import requests
+
+from basis.cli.config import update_local_basis_config
+from basis.cli.services.api import (
+    AuthServer,
+    get_auth_server,
+)
+from basis.cli.services.auth import (
+    LOCAL_OAUTH_PORT,
+    BaseOAuthRequestHandler,
+    execute_oauth_flow,
+)
+
+
+def login():
+    auth_server = get_auth_server()
+
+    code_verifier = base64.urlsafe_b64encode(os.urandom(72)).decode("utf-8")
+    code_verifier = re.sub("[^a-zA-Z0-9]+", "", code_verifier)
```

Comment:
Can you maybe add a comment explaining what's going on here? I see a lot of magic numbers, and at first glance it seems fishy to be b64 encoding some entropy then discarding an unknown amount of it but only for `code_verifier` and not for `state` etc.